[scroll-animations] update effect `phase` and `activeTime` computation to account for percentage values

graouts opened the pull request:
#### 89612b5ede18dd364afff3070f3b09aa04b20211
<pre>
[scroll-animations] update effect `phase` and `activeTime` computation to account for percentage values
<a href="https://bugs.webkit.org/show_bug.cgi?id=281217">https://bugs.webkit.org/show_bug.cgi?id=281217</a>
<a href="https://rdar.apple.com/137668682">rdar://137668682</a>

Reviewed by Simon Fraser.

In order to compute the phases of Scroll-driven Animations we update the `phase` and `activeTime`
computations under `AnimationEffectTiming::getBasicTiming()` to work with percentage time values
by adopting the latest version of their computation procedure as defined by Web Animations Level 2.

Since additional data is needed by `AnimationEffectTiming` to compute those, namely the associated
animation&apos;s associated timeline&apos;s current time and duration, we introduce a new data structure to
pass to `AnimationEffectTiming::getBasicTiming()` and `AnimationEffectTiming::resolve()`.

* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::resolutionData const):
(WebCore::AnimationEffect::getBasicTiming const):
(WebCore::AnimationEffect::getComputedTiming const):
(WebCore::AnimationEffect::localTime const): Deleted.
(WebCore::AnimationEffect::playbackRate const): Deleted.
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/AnimationEffectTiming.cpp:
(WebCore::AnimationEffectTiming::getBasicTiming const):
(WebCore::AnimationEffectTiming::resolve const):
* Source/WebCore/animation/AnimationEffectTiming.h:
* Source/WebCore/animation/CSSNumberishTime.cpp:
(WebCore::CSSNumberishTime::matchingZero const):
(WebCore::CSSNumberishTime::approximatelyLessThan const):
(WebCore::CSSNumberishTime::approximatelyGreaterThan const):
* Source/WebCore/animation/CSSNumberishTime.h:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
(WebCore::AcceleratedEffect::apply):

Canonical link: <a href="https://commits.webkit.org/285040@main">https://commits.webkit.org/285040@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98ca1f42c369856c99afff62eb978bc130ed2a9d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71056 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50468 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23829 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75161 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22261 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73172 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58267 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22079 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56185 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14656 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74122 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61252 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36627 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42512 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18700 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20602 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64436 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19060 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76882 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15288 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18241 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63927 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15332 "Built successfully") | [⏳ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/macOS-Release-WK2-Stress-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63880 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11998 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5634 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10936 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46269 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1045 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47341 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48624 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47083 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/AnimationEffectTiming.h:
```diff
@@ -56,9 +58,17 @@ struct AnimationEffectTiming {
     CSSNumberishTime endTime { 0_s };
 
+    struct ResolutionData {
+        std::optional<CSSNumberishTime> timelineTime;
```
All these optionals make this pretty space-inefficient. Consider `Markable<>`.

graouts replied:
Adding it to my todo list of cleanups.

graouts committed:
[scroll-animations] update effect `phase` and `activeTime` computation to account for percentage values
https://bugs.webkit.org/show_bug.cgi?id=281217
rdar://137668682

Reviewed by Simon Fraser.

In order to compute the phases of Scroll-driven Animations we update the `phase` and `activeTime`
computations under `AnimationEffectTiming::getBasicTiming()` to work with percentage time values
by adopting the latest version of their computation procedure as defined by Web Animations Level 2.

Since additional data is needed by `AnimationEffectTiming` to compute those, namely the associated
animation's associated timeline's current time and duration, we introduce a new data structure to
pass to `AnimationEffectTiming::getBasicTiming()` and `AnimationEffectTiming::resolve()`.

* Source/WebCore/animation/AnimationEffect.cpp:
(WebCore::AnimationEffect::resolutionData const):
(WebCore::AnimationEffect::getBasicTiming const):
(WebCore::AnimationEffect::getComputedTiming const):
(WebCore::AnimationEffect::localTime const): Deleted.
(WebCore::AnimationEffect::playbackRate const): Deleted.
* Source/WebCore/animation/AnimationEffect.h:
* Source/WebCore/animation/AnimationEffectTiming.cpp:
(WebCore::AnimationEffectTiming::getBasicTiming const):
(WebCore::AnimationEffectTiming::resolve const):
* Source/WebCore/animation/AnimationEffectTiming.h:
* Source/WebCore/animation/CSSNumberishTime.cpp:
(WebCore::CSSNumberishTime::matchingZero const):
(WebCore::CSSNumberishTime::approximatelyLessThan const):
(WebCore::CSSNumberishTime::approximatelyGreaterThan const):
* Source/WebCore/animation/CSSNumberishTime.h:
* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
(WebCore::AcceleratedEffect::apply):

Canonical link: https://commits.webkit.org/285040@main
